Added support for Pardiso solver and changed logic for checking MUMPS, Pardiso installs

Pbellive opened the pull request:
Hi Lars,
               I've made a couple of changes to the LinearSolvers.jl submodule. Have a look and merge if you're interested. I added support for Pardiso to LinearSolvers. Pardiso support relies on the [Pardiso.jl](https://github.com/JuliaSparse/Pardiso.jl) package. I also changed how LinearSolvers checks to see if MUMPS and Pardiso are installed. Pkg.installed() doesn't work on remote workers running on different nodes than the master process (see JuliaLang/julia#16222). Because of that limitation, LinearSolvers.jl throws an error when trying to load on multiple cluster nodes. My solution is to simply put the loading of the MUMPS and Pardiso modules in try/catch blocks:

```
hasMUMPS = false
try
  using MUMPS
  hasMUMPS = true
catch
end
```

The pull request also includes a small and very rudimentary wrapper of `projGNCG` for doing iteratedTikhonov to adjust the regularization parameter over the course of the inversion. This was a quick and dirty function I added for a specific project I was working on. If you're interested in merging the LinearSolvers changes but not adding the iteratedTikhonov wrapper let me know and I'll clean up the PR.

Patrick

Pbellive committed:
Added support for Pardiso solver to LinearSolvers submodule, added iteratedTikhonov wrapper of projGNCG. Also a few minor bug fixes.
Changed logic that checks if MUMPS and Pardiso are installed to avoid calling Pkg.installed(). Pkg. calls do not work on remote workers (See JuliaLang issue 16222). This was causing jInv.linearSolvers to crash when run over multiple cluster nodes.

lruthotto commented on src/InverseSolve/regularizers.jl:
```diff
@@ -124,6 +124,50 @@ function wdiffusionReg(m::Vector, mref::Vector, M::AbstractMesh; Iact=1.0, C=[])
    return Rc,dR,d2R
 end # function wdiffusionReg
 
+function wdiffusionAndDepthReg(m::Vector, mref::Vector, M::AbstractMesh; Iact=1.0, C=[])
```
It looks like this function combines `wdiffusionReg` and `smallnessReg`, right? If so, we don't need to add a new function for this. In `pInv`  the regularization function, parameter, and reference model can be stacked into arrays. But I may overlook something here.

lruthotto commented on src/InverseSolve/iteratedTikhonov.jl:
```diff
@@ -0,0 +1,51 @@
+    for i=1:length(DcRef)
+      DcLoc = [DcLoc;vec(fetch(DcRef[i]))]
+    end
+    F,dF,d2F = pInv.misfit(DcLoc,dObs,Wd)
```
the current misfit `F` should be stored in `hist`. If you removed this line here, it seems to me that you don't need to add two methods for `iteratedTikhonov`.

lruthotto commented on test/jInv.out:
```diff
@@ -0,0 +1,6 @@
+i.LS	       F	       R	alpha[1]	   Jc/J0	 #Active
```
please remove this file from PR

lruthotto commented on src/InverseSolve/iteratedTikhonov.jl:
```diff
@@ -0,0 +1,51 @@
+export iteratedTikhonov
+
+function iteratedTikhonov(m0,pInv::InverseParam,PF::Array{RemoteRef{Channel{Any}}},nBeta,betaFac,targetMisfit;indCredit=[],dumpResults::Function = dummy)
```
please provide some documentation for the function so that

```
?iteratedTikhonov
```

works. Look, for example, at `projGNCG`.

lruthotto commented on src/InverseSolve/iteratedTikhonov.jl:
```diff
@@ -0,0 +1,51 @@
+export iteratedTikhonov
```
Can you please add a minimal example for `iteratedTikhonov` and include it into `test/runtests.jl`. You should be able to modify `test/InverseSolve/testLeastSquares.jl`

Pbellive commented:
Hi Lars,
               Thanks for the comments and sorry for the quite sloppy PR. I forgot to clear out junk from my local repo before pushing. Re wdiffusionAndDepthReg function: this was part of an aborted attempt to add depth weighted regularization. I've removed the function from regularizers.jl. 

Re iteratedTikhonov.jl, I'll clean it up, add documentation and tests, then update the PR.

Pbellive commented:
Hi Lars,
              I cleaned up the iteratedTikhonov function. It now has a single, much simpler method and the output has been improved. I added documentation (`?iteratedTikhonov` works) and added tests to .../test/InverseSolve/testLeastSquares.jl.

I also fixed what I thought were small bugs in projGNCG and barrierGNCG. A message that GNCG did not converge after `maxIter` iterations was being output when convergence occured on the `maxIter`ith iteration. If that was the intended behaviour and not a bug, my apologies.

lruthotto commented:
Thanks Patrick. I think we are ready to merge this one. Can you please squash your commits into one to keep the number of versions small? See https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit for reference. Thanks.

Pbellive commented:
Hi Lars,
             All the changes have now been squashed into a single commit whose parent is the current master.

lruthotto commented:
awesome! Thanks, Patrick
